st2common/st2common/util/param.py

```diff
@@ -174,13 +174,22 @@ def _validate(G):
     """
     Validates dependency graph to ensure it has no missing or cyclic dependencies
     """
+    g_copy = G.copy()
     for name in G.nodes:
```

Comment:
These are the nodes I'm talking about. Each "node" in the graph represents a named var in a string.

Within a single string, either all of the variables (nodes) should be valid, or none of them. If none of them are valid, then pass the string as-is.

You mentioned parameters, but I'm only talking about a single string, like the default value of a parameter. If that string does not have any valid cars, then assume it is not actually a Jinja template.

Reply:
It makes sense now. Thanks for the explanation.

Reply:
dependencies are added as nodes when it does `meta.find_undeclared_variables` and `G.add_edge(dependency, name)`.  At this point the nodes are all in the graph.

Reply:
each root node should represent an original parameter and the edges should point to jinja template variable nodes

Reply:
each node represents all vars including parameters.  not just named vars in strings.

Reply:
In order to do what you ask the `_resolve_dependencies` would need to instead perform the removal of the invalid edges and nodes if all 'predecessors' or 'children' are invalid. (yes it is backwards)


@nzlosh the problem with raw_string option is it doesn't solve objects that have embedded jinja in them.

Reply:
or I could intelligently process the root nodes first then iterate through children with recursion in `_validate`.  currently it just iterates through nodes without any order.

Reply:
I did it, but I don't really like it.  I ended up having to add a lot of complexity.

Reply:
I still need to add a unit test for the raised exception in the case of some but not all dependent variables exist.  Also, I have no idea if black or pylint will complain.